src/main/scala/com/gu/thrift/serializer/ThriftDeserializer.scala

```diff
@@ -8,30 +8,32 @@ import org.apache.thrift.transport.TIOStreamTransport
 import scala.util.Try
 
 object ThriftDeserializer {
-
-  /** Reads a Thrift value from a byte array. By default, will try to read
+  /** Reads a Thrift value from a byte buffer. By default, will try to read
     * the first byte for details on how the value is encoded. Plain
     * values can be decoded by setting the `noHeader` flag to `true`.
     */
-  def deserialize[T <: ThriftStruct : ThriftStructCodec](buffer: Array[Byte], noHeader: Boolean): Try[T] =
-    Try {
-      if (!noHeader) {
-        val settings = buffer.head
-        val compressionType = compression(settings)
-        compressionType match {
-          case NoneType => payload(buffer.tail)
-          case GzipType => payload(GzipCompression.uncompress(buffer.tail))
-          case ZstdType => payload(ZstdCompression.uncompress(buffer.tail))
-        }
-      } else payload(buffer)
+  def deserialize[T <: ThriftStruct : ThriftStructCodec](buffer: ByteBuffer, noHeader: Boolean): Try[T] = Try {
+    if(!noHeader) {
+      val settings = buffer.get() //also increments buffer position by 1, so buffer.slice() below returns the "tail"
+      val compressionType = compression(settings)
+      compressionType match {
+        case NoneType => payload(buffer.slice())
+        case GzipType => payload(GzipCompression.uncompress(buffer.slice()))
+        case ZstdType => payload(ZstdCompression.uncompress(buffer.slice()))
+      }
+    } else {
+      payload(buffer)
     }
+  }
 
-  /** Reads a Thrift value from a byte array. Tries to read the first byte
-    * for details on how the value is encoded and decode accordingly, or
-    * decode as a plain value if it fails.
+  /** Reads a Thrift value from a byte array. By default, will try to read
+    * the first byte for details on how the value is encoded. Plain
+    * values can be decoded by setting the `noHeader` flag to `true`.
     */
-  def deserialize[T <: ThriftStruct : ThriftStructCodec](buffer: Array[Byte]): Try[T] =
-    deserialize(buffer, false) orElse deserialize(buffer, true)
+  def deserialize[T <: ThriftStruct : ThriftStructCodec](buffer: Array[Byte], noHeader: Boolean): Try[T] = deserialize(ByteBuffer.wrap(buffer), noHeader)
+
+  def deserialize[T <: ThriftStruct : ThriftStructCodec](buffer: Array[Byte]):Try[T] = deserialize(buffer, false) orElse deserialize(buffer, true)
+  def deserialize[T <: ThriftStruct : ThriftStructCodec](buffer: ByteBuffer):Try[T] = deserialize(buffer, false) orElse deserialize(buffer, true)
```

Comment:
This threw me for a minute - it looked as though ByteBuffer and Array[Byte] were treated as interchangeable, but we know they're not, and really it's the definition on line 33 that's wrapping the Array in a ByteBuffer! Fun times.